Forbid `PretrainedConfig` from saving `generate` parameters; Update deprecations in `generate`-related code 🧹

gante opened the pull request:
# What does this PR do?

See title :) I did a quick scan and update over things that fulfilled the following criteria:
- had deprecation messages
- were related to what I've been working on

______________________________________________

Key changes: 
1. Saving a model config with parameters meant to be in the generation config is no longer allowed
2. However, at model saving time (which saves related configs), move those parameters to the generation config before potentially complaining about it. This is to prevent fine-tuning jobs from failing after significant compute $ is spent.

gante committed:
deprecated stuff

gante commented on src/transformers/commands/pt_to_tf.py:
```diff
@@ -12,45 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import inspect
```
TL;DR disabled TF conversion (the warning said it would be removed in v4.43), but kept the error message which redirects to the safetensor conversion guide :)

gante commented on src/transformers/modeling_utils.py:
```diff
@@ -2551,24 +2551,6 @@ def save_pretrained(
             if not _hf_peft_config_loaded:
                 model_to_save.config.save_pretrained(save_directory)
             if self.can_generate():
-                # generation config built from the model config + the model config holds generation kwargs -> generate
```
NOTE: the check removed here is redundant with the check that gets upgraded from warning to exception in this PR. We call `config.save_pretrained()` above in L2552, which runs the upgraded check 👼

gante commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -693,7 +693,7 @@ def forward(
             logger.warning_once(
                 "Using `past_key_values` as a tuple is deprecated and will be removed in v4.45. "
-                "Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/v4.41.3/en/internal/generation_utils#transformers.Cache)"
+                "Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/internal/generation_utils#transformers.Cache)"
```
found in my quick scan for deprecations (CMD+F on "v4.41") -- let's use our latest docs to guide our users

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_32659). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante committed:
update test

gante committed:
update test

amyeroberts reviewed (approved):
Killing old code 🤩

gante committed:
automagically fix the bad configs

gante commented on src/transformers/configuration_utils.py:
```diff
@@ -1078,14 +1075,29 @@ def _get_generation_defaults() -> Dict[str, Any]:
+                # 1. The parameter is set to a default generation value (from `generate`'s perspective, it's the same
+                # as if nothing is set)
+                is_default_generation_value = getattr(self, parameter_name) == default_value
+                # 2. The parameter is set as default in the model config (BC support for models like BART)
```
This situation was not encoded before this PR (and caused tests to fail, god bless tests 🙏 )

gante commented on src/transformers/modeling_utils.py:
```diff
@@ -2549,26 +2549,20 @@ def save_pretrained(
         # Save the config
         if is_main_process:
             if not _hf_peft_config_loaded:
+                # If the model config has set attributes that should be in the generation config, move them there.
```
Instead of warning (before this PR) / raising an exception (this PR if L2552-2563 were not added) when `model.config` holds things that should be in `model.generation_config` AND were explicitly set by the user (e.g. `model.config.top_k = 5`), now we automagically move misplaced items ✨ 

Automagically moving things is not usually recommended, but this is a special case: `model.save_pretrained` is often run at the end of training jobs, we really want to avoid expensive crashes

gante commented on src/transformers/models/llama/modeling_llama.py:
```diff
@@ -957,7 +957,7 @@ def forward(
             past_key_values = DynamicCache.from_legacy_cache(past_key_values)
             logger.warning_once(
                 "We detected that you are passing `past_key_values` as a tuple and this is deprecated and will be removed in v4.43. "
-                "Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/v4.41.3/en/internal/generation_utils#transformers.Cache)"
+                "Please use an appropriate `Cache` class (https://huggingface.co/docs/transformers/internal/generation_utils#transformers.Cache)"
             )
```
We may want to use `DynamicCache.from_legacy_cache` in non-deprecated ways :p for instance, some optimum integrations use it

gante committed:
fix test

gante committed:
handle the case of composite model configs (which have no defaults)

gante committed:
nit

gante committed:
boolean should be True

gante committed:
commenting logic

gante committed:
improve logic (default config should be the first thing we check)

gante committed:
whisper - remove custom generative config attribute

gante commented:
(@amyeroberts don't review yet, I'm going through failed tests -- mostly cases where we are explicitly setting `generate` arguments in `model.config`, precisely what we don't want users to do)

gante committed:
tf/flax whisper -- remove unwanted custom config

gante committed:
remove custom bart config

gante committed:
raise ValueError instead of assert

gante committed:
remove test for deleted feature

gante committed:
default config preference

gante committed:
yup, that should raise a warning

gante committed:
another model tester with generation params

gante committed:
another model

gante committed:
single worker for tf examples job

gante committed:
handle defaults for composite models

gante committed:
update tf whisper logits processors to handle edge case

gante committed:
even more precise

gante committed:
move defaults up

gante committed:
last one?

gante committed:
handle exception

gante committed:
one more edge case

gante committed:
another one

amyeroberts commented:
@gante I'm going to unsubscribe to avoid notifications for each push. Just ping / @ me to let me know when it's ready for review again!

gante committed:
this one is flaky

gante commented on .circleci/create_circleci_config.py:
```diff
@@ -327,7 +327,8 @@ def job_name(self):
     install_steps=["uv venv && uv pip install . && uv pip install -r examples/tensorflow/_tests_requirements.txt"],
-    parallelism=8
+    parallelism=8,
+    pytest_num_workers=1
```
(the default is 12 workers)
Without this, we're having a concurrency error downloading NLTK data ([e.g.](https://app.circleci.com/pipelines/github/huggingface/transformers/100691/workflows/56c6d291-9d7a-481d-8e29-dcbf96e2bc5d/jobs/1341285)). If we track down that exception, we end up in the NLTK repo, where a maintainer suggests not having concurrency as a fix

gante replied:
NTLK issue: https://github.com/nltk/nltk/issues/3248 

Workarounds would all consist of running this part of the CI job without parallelism or adding locks -- IMO non-trivial stuff. Given that it blocks any PR that touches code that might be used in TF, such as `configuration_utils.py`, it's pretty disruptive 😢 I think adding this band-aid and chasing a better solution in parallel would be the best 🤗

ydshieh replied:
I don't know what happen recently, but we don't have this error before.

If we decide to have this quick (temp) fix, it might be a good idea to have a small PR for only this part.

gante replied:
Opening a separate PR :)

gante commented on src/transformers/configuration_utils.py:
```diff
@@ -81,6 +81,15 @@ class PretrainedConfig(PushToHubMixin):
       model.
     - **num_hidden_layers** (`int`) -- The number of blocks in the model.
 
+    <Tip warning={true}>
+
+    Setting parameters for sequence generation in the model config is deprecated. For backward compatibility, loading
+    some of them will still be possible, but attempting to overwrite them will throw an exception -- you should set
+    them in a [~transformers.GenerationConfig]. Check the documentation of [~transformers.GenerationConfig] for more
+    information about the individual parameters.
+
+    </Tip>
```
Instead of documenting `generate` parameters in `PretrainedConfig`, which is deprecated, show a warning in the docs pointing towards `GenerationConfig`

gante commented on src/transformers/generation/tf_logits_process.py:
```diff
@@ -520,15 +520,21 @@ def __init__(self, begin_suppress_tokens, begin_index):
-            ),
-            lambda: scores,
-        )
+        suppressed_indices = []
+        for token in self.begin_suppress_tokens:
+            if token < scores.shape[-1]:  # to ensure we don't go beyond the vocab size
+                suppressed_indices.extend([[i, token] for i in range(scores.shape[0])])
```
We disallowed storing custom generate flags in the model config -> we can no longer set custom generate parameterization in the model tester [which is not necessarily bad, all models should be able to generate with their default parameters!] -> we couldn't hardcode our way out of edge cases :)

This change fixes the edge case where we set a token that's not in the vocab to be suppressed. The alternative would be to reconfigure the tester to have a vocab size larger than the default value for this parameter, >50k tokens. Given that it is pretty harmless, I went with this :)

gante replied:
@amyeroberts We can increase the vocabulary size of the tester to the original vocabulary size of 50+k tokens (instead of the dummy vocab size of 100).

Happy to do that instead :)

[Note: given TF's indexing, where GPU+no XLA gladly accepts out-of-bounds indexes, this logic actually streamlines the behavior across devices and/or XLA.]

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1359,19 +1354,18 @@ def _prepare_generation_config(
                 not is_torchdynamo_compiling()
-                and self.generation_config._from_model_config
-                and self.generation_config._original_object_hash == hash(self.generation_config)
-                and self.config._has_non_default_generation_parameters()
```
this was actually incorrect: we want to detect changes against the default model config, not against the default generation parameters

(To recap: if a user updates the model config and the generation config was automatically generated from it, throw a warning)

gante commented on src/transformers/models/encoder_decoder/configuration_encoder_decoder.py:
```diff
@@ -74,9 +74,11 @@ class EncoderDecoderConfig(PretrainedConfig):
-            "encoder" in kwargs and "decoder" in kwargs
-        ), "Config has to be initialized with encoder and decoder config"
+        if "encoder" not in kwargs or "decoder" not in kwargs:
+            raise ValueError(
```
I added a try/except for composite models, catching `ValueError`, which led to to this discovery :) You'll see this pattern fixed in other composite models.

gante commented on tests/models/bart/test_modeling_bart.py:
```diff
@@ -123,12 +123,6 @@ def __init__(
-        # if however the token to be generated is already at -inf then it can lead token
-        # `nan` values and thus break generation
-        self.forced_bos_token_id = None
-        self.forced_eos_token_id = None
```
If we go check the PR that added these, we can read that this change was added to maybe fix a bug that happens sporadically 👀 

I don't agree with the description of the problem given our current code base, I believe the root cause has been addressed. Furthermore, it's good to remove fudge factors 😉

gante commented on tests/models/blenderbot/test_modeling_blenderbot.py:
```diff
@@ -368,7 +360,6 @@ def __init__(
         decoder_attention_heads=4,
         max_position_embeddings=30,
         is_encoder_decoder=False,
-        encoder_no_repeat_ngram_size=0,
```
no clue why this would be here 🤷

gante commented on tests/models/decision_transformer/test_modeling_decision_transformer.py:
```diff
@@ -41,7 +41,6 @@ def __init__(
         act_dim=6,
         state_dim=17,
         hidden_size=23,
-        max_length=11,
```
No idea why this was here :)

gante commented on tests/models/mamba2/test_modeling_mamba2.py:
```diff
@@ -227,6 +227,10 @@ def test_multi_gpu_data_parallel_forward(self):
     def test_generate_from_inputs_embeds_decoder_only(self):
         pass
 
+    @unittest.skip(reason="To fix, Mamba 2 cache slicing test case is an edge case")
+    def test_inputs_embeds_matches_input_ids_with_generate(self):
+        pass
```
this test was failing on `main` -- perhaps because of concurrent merges?

gante commented on tests/models/recurrent_gemma/test_modeling_recurrent_gemma.py:
```diff
@@ -413,6 +413,10 @@ def _check_hidden_states_for_generate(
     def test_initialization(self):
         pass
 
+    @unittest.skip(reason="TODO @gante not super important and failing.")
+    def test_inputs_embeds_matches_input_ids_with_generate(self):
+        pass
```
Failing on `main` too.

I suspect this didn't get detected in the (recent) PR that added it because of our test fetched, which only checks the main models on these wide changes (?)

gante replied:
https://github.com/huggingface/transformers/issues/32913

(The issue catches a wider net than skipping `test_inputs_embeds_matches_input_ids_with_generate`, as this skip is merely a symptom of the root problem :) )

gante commented:
@amyeroberts Ready 🙌 

Notes:
1. saving a config with custom generate parameters now creates an exception -> all testers that relied on setting custom generation parameters to work around edge cases failed [because we pass the tester args to the model config] -> this PR fixes edge cases/updates tests 💪 
2. touching many files triggers a lot of test runs. A few unrelated broken tests were found (and fixed) in the process
3. I've added GH comments on details that are both not useful for the future codebase AND not obvious why the change was made
4. an unrelated pipeline test is failing due to timeout, triggering a new run is not making CI green. I'll need you to merge when you're happy with the PR :D

gante committed:
make fixup

amyeroberts commented on src/transformers/models/audio_spectrogram_transformer/configuration_audio_spectrogram_transformer.py:
```diff
@@ -118,3 +120,9 @@ def __init__(
         self.max_length = max_length
         self.num_mel_bins = num_mel_bins
+
+    # Overwritten from the parent class: AST is not compatible with `generate`, but has a config parameter sharing the
+    # same name (`max_length`). Sharing the same name triggers checks regarding the config -> generation_config
+    # generative parameters deprecation cycle, overwriting this function prevents this from happening.
+    def _get_non_default_generation_parameters(self) -> Dict[str, Any]:
+        return {}
```
I'm not sure I understand this - why would there be checks run on a config -> generation_config if the model isn't compatible with generate? 

It seems like a bad pattern to need a generate specific method added to config if the model can't be used for generation

gante replied:
I understand and share your pain 😬 Allow me to elaborate.

The root issue is in the model config. A model config is agnostic to whether the model can or cannot generate, unless we hardcode it in the config class or access external objects. 

When we call `config.save_pretrained`, we don't know whether we should or should not check generation parameters. Since we want to move all generation-related logic to `GenerationConfig`, whose one of the goals is to avoid clashes like this and allow proper validation, we have to assume all configs may be holding generate parameters. The current logic enforces no custom generate parameters are held in the model config, so no new models parameterize `generate` through the model config.

AST is the exception. It doesn't support `generate`, but it has a model parameter whose name is also used in `generate` (`max_length`). If a user creates a model with a custom `max_length`, which should be allowed, the current `save_pretrained` would crash because it thinks it has a custom `generate` parameter. This workaround effectively skips the `generate` checks, since we know it can't generate.

Between all three alternatives I could see, this seemed the best solution -- especially since it is a temporary fix, to be removed after we untangle the two configs (v5.0?)

Considered fixes:
- skip generate-related checks on this specific class (implemented)
- add a parameter to ALL model config classes to store whether the model can generate
- add an argument to `config.save_pretrained` to check whether to check `generate` parameters, receive `model.can_generate()` when called from `model.save_pretrained`

amyeroberts replied:
Make sense - thanks for taking the time to explain!

amyeroberts reviewed (commented):
Thanks for handling all of these updates! 

A few questions / comments but overall looks good 🤗

gante committed:
accept storing None; moving sets to None

gante commented on tests/utils/test_modeling_utils.py:
```diff
@@ -1599,14 +1599,17 @@ def test_safetensors_torch_from_torch_sharded(self):
-            self.assertEqual(len(logs.output), 1)
-            self.assertIn("Your generation config was originally created from the model config", logs.output[0])
+            model.save_pretrained(tmp_dir)
+            self.assertTrue(model.config.repetition_penalty != 3.0)
```
@amyeroberts that is a great question -- what gets written in `model.config` in this case. I agree with your comment (`If it's its old value, I'm worried this could cause confusion`), which was the actual behavior in the commit you reviewed, so I've made the following adjustments:
1. Moving a parameter from `model.config` to `model.generation_config` sets that parameter to `None` in `model.config`.
2. Storing a generation parameter as `None` in `model.config` is now accepted (even if it is not the default value);
3. Updated the warning message, explaining to the user why they are seeing it
4. The test now also checks that the warning is thrown. It is also commented, so it should be simple to understand what should be happening.
🤗

gante committed:
Update src/transformers/models/rag/configuration_rag.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

gante committed:
Update src/transformers/models/encoder_decoder/configuration_encoder_decoder.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

gante commented:
(CI was failing, rebased with main to check if that was the issue)

gante committed:
remove CI TF examples fix (moved to #32935)

amyeroberts commented on tests/utils/test_modeling_utils.py:
```diff
@@ -1599,14 +1600,30 @@ def test_safetensors_torch_from_torch_sharded(self):
             self.assertTrue(torch.equal(p1, p2))
 
-    def test_modifying_model_config_causes_warning_saving_generation_config(self):
+    def test_modifying_model_config_gets_moved_to_generation_config(self):
```
Nice :)

amyeroberts reviewed (approved):
Thanks for iterating and taking the time to answer all my qs. 

As this touches so many models and core logic, we should ideally do a `[run_all]` run of slow tests to make sure everything is tip-top before merge

gante committed:
Merge branch 'main' into deprecations_aug

gante committed:
[test_all] trigger all tests
